### ISSUE TITLE
Type stability broken on 0.6

### ISSUE THREAD

**Issue by KristofferC** [JuliaArrays/StaticArrays.jl#88]
Currently, type stability for basic operation is broken on julia 0.6. This is due to https://github.com/JuliaLang/julia/pull/19421.

```
julia> code_warntype(*, Tuple{SVector{3,Float64}, Float64})
Variables:
  #self#::Base.#*
  a::StaticArrays.SVector{3,Float64}
  b::Float64
  #201::StaticArrays.##201#202{Base.#*,Float64}
  newtype::Any
  T::Any
  #temp#@_7::Any
  #temp#@_8::Any
  #temp#@_9::Any
```

The issue does not require any immediate action here because 0.6 is in constant fluctuation, but I thought it would be good to have it tracked.
Also, perhaps there should be some `@inferred` tests? I got aware of this from https://github.com/KristofferC/ContMechTensors.jl/blob/593ce54b36cf25e83ee8b32459b4dccb1d643797/test/test_misc.jl#L132.

**Comment by andyferris** [comment]
I wasn't planning on playing with anything in v0.6 until Jeff's subtype branch is merged...

Good idea about adding more inference tests, but we should keep in mind that `@inferred` doesn't catch problems with element type (you might notice there are some type assertions for that).

**Comment by c42f** [comment]
This appears to have been fixed now in 0.6, or at least the code for that particular function looks ok on my build from today.
